README.md
# Keccak-f[1600] NIF (SHA3-224, SHA3-256, SHA3-384, SHA3-512, SHAKE128, SHAKE256, SPONGERNG)

[![Build Status](https://travis-ci.org/potatosalad/erlang-keccakf1600.svg?branch=master)](https://travis-ci.org/potatosalad/erlang-keccakf1600) [![Hex.pm](https://img.shields.io/hexpm/v/keccakf1600.svg)](https://hex.pm/packages/keccakf1600)

[Keccak-f[1600]](http://keccak.noekeon.org/) NIF with timeslice reductions for Erlang and Elixir.

The timeslice reductions allow the NIF to perform operations on very large inputs without blocking the scheduler or requiring the Erlang VM to support dirty schedulers.  See the [bitwise](https://github.com/vinoski/bitwise) project from which the strategy was derived.

Tested against the [FIPS 202](http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf) test vectors.

## Installation

Add `keccakf1600` to your project's dependencies in `mix.exs`

```elixir
defp deps do
  [
    {:keccakf1600, "~> 3.0.0"}
  ]
end
```

Add `keccakf1600` to your project's dependencies in your `Makefile` for [`erlang.mk`](https://github.com/ninenines/erlang.mk) or the following to your `rebar.config`

```erlang
{deps, [
  {keccakf1600, ".*", {git, "git://github.com/potatosalad/erlang-keccakf1600.git", {branch, "master"}}}
]}.
```

## Usage

This library follows usage semantics from Erlang's own [`crypto`](http://erlang.org/doc/man/crypto.html) library, with the exception of the SHAKE128, SHAKE256, and SPONGERNG algorithms as described below.

##### SPONGERNG

#### `keccakf1600_spongerng:init_from_buffer/2`

This function allows you to specify an initial seed buffer and whether the PRNG will be deterministic or not.

```erlang
%% Deterministic
keccakf1600_spongerng:init_from_buffer(<<>>, true).
% {spongerng, #Ref<0.0.0.1>}
%% Non-deterministic
keccakf1600_spongerng:init_from_buffer(<<>>, false).
% {spongerng, #Ref<0.0.0.2>}
```

#### `keccakf1600_spongerng:init_from_file/3`

This function allows you specify an initial seed file up to the given length and whether the PRNG will be deterministic or not.

```erlang
%% Deterministic
keccakf1600_spongerng:init_from_file("seed.txt", 16, true).
% {spongerng, #Ref<0.0.0.3>}
%% Non-deterministic
keccakf1600_spongerng:init_from_file("seed.txt", 16, false).
% {spongerng, #Ref<0.0.0.4>}
```

#### `keccakf1600_spongerng:init_from_dev_urandom/0`

This function reads an initial seed from `/dev/urandom` and is only allowed to be non-deterministic.

```erlang
%% Non-deterministic
keccakf1600_spongerng:init_from_dev_urandom().
% {spongerng, #Ref<0.0.0.5>}
```

#### `keccakf1600_spongerng:next/2`

This function returns the next length of bytes from the sponge and returns the new sponge state.

```erlang
Sponge0 = keccakf1600_spongerng:init_from_buffer(<<>>, true),
{Sponge1, Output} = keccakf1600_spongerng:next(Sponge0, 8).
% {{spongerng, #Ref<0.0.0.6>}, <<99,190,253,62,125,162,80,150>>}
```

#### `keccakf1600_spongerng:stir/2`

This function modifies the sponge state (stirs the pot) with the given input.

```erlang
Sponge0 = keccakf1600_spongerng:init_from_buffer(<<>>, true),
Sponge1 = keccakf1600_spongerng:stir(Sponge0, <<"test">>),
{Sponge2, Output} = keccakf1600_spongerng:next(Sponge1, 8).
% {{spongerng, #Ref<0.0.0.7>}, <<168,214,5,0,60,110,186,33>>}
```

### SHA-3

#### `keccakf1600_sha3:hash/2`

This function can be used for the following algorithms:

 * SHA3-224 (`sha3_224`)
 * SHA3-256 (`sha3_256`)
 * SHA3-384 (`sha3_384`)
 * SHA3-512 (`sha3_512`)

```erlang
keccakf1600_sha3:hash(sha3_224, <<"test">>).
% <<55,151,191,10,251,191,202,74,123,187,167,96,42,43,85,39,70,135,101,23,167,249,183,206,45,176,174,123>>

keccakf1600_sha3:hash(sha3_256, <<"test">>).
% <<54,240,40,88,11,176,44,200,39,42,154,2,15,66,0,227,70,226,118,174,102,78,69,238,128,116,85,116,226,245,171,128>>

keccakf1600_sha3:hash(sha3_384, <<"test">>).
% <<229,22,218,187,35,182,227,0,38,134,53,67,40,39,128,163,174,13,204,240,85,81,207,2,149,23,141,127,240,241,180,30,236,185,219,63,242,25,0,124,78,9,114,96,213,134,33,189>>

keccakf1600_sha3:hash(sha3_512, <<"test">>).

CHANGELOG.md
# Changelog

## 3.0.0 (2019-07-28)

* Enhancements
  * Refactor to match [erlang-libdecaf](https://github.com/potatosalad/erlang-libdecaf).
  * Add support for SPONGERNG functions.
  * Improved testing setup and hopefully fix build problems on FreeBSD.

## 2.0.0 (2016-02-05)

* Enhancements
  * Rewrite as NIF with timeslice reductions (roughly 10x faster).

* Breaking Changes
  * Previous API has been simplified to the following (using `sha3_512` as an example):
    * `Out = keccakf1600:hash(sha3_512, <<"test">>)`
    * `State0 = keccakf1600:init(sha3_512)`
    * `State1 = keccakf1600:update(State0, <<"test">>)`
    * `Out = keccakf1600:final(State1)`
  * The `shake128` and `shake256` algorithms are very similar, but require an output length (using `shake256` and `64` output length as an example):
    * `Out = keccakf1600:hash(shake256, <<"test">>, 64)`
    * `State0 = keccakf1600:init(shake256)`
    * `State1 = keccakf1600:update(State0, <<"test">>)`
    * `Out = keccakf1600:final(State1, 64)`

## 1.0.2 (2016-02-04)

* Fixes
  * Minor rebar configuration fix.

## 1.0.1 (2016-02-04)

* Fixes
  * Minor rebar configuration fix.

## 1.0.0 (2016-02-04)

* Enhancements
  * Streaming support
    * `keccakf1600_sha3_224:init/0`, `keccakf1600_sha3_224:update/2`, `keccakf1600_sha3_224:final/1`
    * `keccakf1600_sha3_256:init/0`, `keccakf1600_sha3_256:update/2`, `keccakf1600_sha3_256:final/1`
    * `keccakf1600_sha3_384:init/0`, `keccakf1600_sha3_384:update/2`, `keccakf1600_sha3_384:final/1`
    * `keccakf1600_sha3_512:init/0`, `keccakf1600_sha3_512:update/2`, `keccakf1600_sha3_512:final/1`
    * `keccakf1600_shake128:init/0`, `keccakf1600_shake128:update/2`, `keccakf1600_shake128:final/2`
    * `keccakf1600_shake256:init/0`, `keccakf1600_shake256:update/2`, `keccakf1600_shake256:final/2`

## 0.0.1 (2016-01-20)

* Initial Release

* Publish to [hex.pm](https://hex.pm/packages/keccakf1600).

* Library Support
  * `keccakf1600_fip202:shake128/2`
  * `keccakf1600_fip202:shake256/2`

c_src/keccakf1600_nif.c
// -*- mode: c; tab-width: 8; indent-tabs-mode: 1; st-rulers: [70] -*-
// vim: ts=8 sw=8 ft=c noet

#include "keccakf1600_nif.h"
#include "shake.h"

/*
 * Erlang NIF functions
 */

#define SHA3_NIF(bits, bytes)	\
	static ERL_NIF_TERM	\
	keccakf1600_sha3_##bits##_nif_1(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		ErlNifBinary in;	\
	\
		if (argc != 1 || !enif_inspect_binary(env, argv[0], &in)) {	\
			return enif_make_badarg(env);	\
		}	\
	\
		if (in.size <= MAX_PER_SLICE) {	\
			ERL_NIF_TERM out;	\
			unsigned char *buf = enif_make_new_binary(env, bytes, &out);	\
	\
			(void) sha3_##bits##_hash(buf, bytes, in.data, in.size);	\
	\
			return out;	\
		}	\
	\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource = enif_alloc_resource(resource_type, sizeof(sha3_##bits##_ctx_t));	\
		struct sha3_##bits##_ctx_s *sponge = (struct sha3_##bits##_ctx_s *)(resource);	\
		(void) sha3_##bits##_init(sponge);	\
	\
		ERL_NIF_TERM newargv[4];	\
	\
		newargv[0] = argv[0];					/* In */	\
		newargv[1] = enif_make_ulong(env, MAX_PER_SLICE);	/* MaxPerSlice */	\
		newargv[2] = enif_make_ulong(env, 0);			/* Offset */	\
		newargv[3] = enif_make_resource(env, resource);		/* Sponge */	\
	\
		(void) enif_release_resource(resource);	\
	\
		return enif_schedule_nif(env, "sha3_" #bits, 0, keccakf1600_sha3_##bits##_nif_4, 4, newargv);	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_sha3_##bits##_nif_4(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		ErlNifBinary in;	\
		unsigned long max_per_slice;	\
		unsigned long offset;	\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource;	\
	\
		if (argc != 4 || !enif_inspect_binary(env, argv[0], &in)	\
				|| !enif_get_ulong(env, argv[1], &max_per_slice)	\
				|| !enif_get_ulong(env, argv[2], &offset)	\
				|| !enif_get_resource(env, argv[3], resource_type, &resource)) {	\
			return enif_make_badarg(env);	\
		}	\
	\
		struct sha3_##bits##_ctx_s *sponge = (struct sha3_##bits##_ctx_s *)(resource);	\
	\
		struct timeval start;	\
		struct timeval stop;	\
		struct timeval slice;	\
		unsigned long end;	\
		unsigned long i;	\
		int percent;	\
		int total = 0;	\
	\
		end = offset + max_per_slice;	\
	\
		if (end > in.size) {	\
			end = in.size;	\
		}	\
	\
		i = offset;	\
	\
		while (i < in.size) {	\
			(void) gettimeofday(&start, NULL);	\
			(void) sha3_##bits##_update(sponge, (uint8_t *)(in.data) + i, end - i);	\
			i = end;	\
			if (i == in.size) {	\
				break;	\
			}	\
			(void) gettimeofday(&stop, NULL);	\
			/* determine how much of the timeslice was used */	\
			timersub(&stop, &start, &slice);	\
			percent = (int)((slice.tv_sec*1000000+slice.tv_usec)/10);	\
			total += percent;	\
			if (percent > 100) {	\
				percent = 100;	\
			} else if (percent == 0) {	\
				percent = 1;	\
			}	\
			if (enif_consume_timeslice(env, percent)) {	\
				/* the timeslice has been used up, so adjust our max_per_slice byte count based on the processing we've done, then reschedule to run again */	\
				max_per_slice = i - offset;	\
				if (total > 100) {	\
					int m = (int)(total/100);	\
					if (m == 1) {	\
						max_per_slice -= (unsigned long)(max_per_slice*(total-100)/100);	\
					} else {	\
						max_per_slice = (unsigned long)(max_per_slice/m);	\
					}	\
				}	\
				ERL_NIF_TERM newargv[4];	\
				newargv[0] = argv[0];					/* In */	\
				newargv[1] = enif_make_ulong(env, max_per_slice);	/* MaxPerSlice */	\
				newargv[2] = enif_make_ulong(env, i);			/* Offset */	\
				newargv[3] = argv[3];					/* Sponge */	\
				return enif_schedule_nif(env, "sha3_" #bits, 0, keccakf1600_sha3_##bits##_nif_4, argc, newargv);	\
			}	\
			end += max_per_slice;	\
			if (end > in.size) {	\
				end = in.size;	\
			}	\
		}	\
	\
		ERL_NIF_TERM out;	\
		unsigned char *buf = enif_make_new_binary(env, bytes, &out);	\
	\
		(void) sha3_##bits##_final(sponge, buf, bytes);	\
		(void) sha3_##bits##_destroy(sponge);	\
	\
		return out;	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_sha3_##bits##_init_nif_0(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		if (argc != 0) {	\
			return enif_make_badarg(env);	\
		}	\
	\
		ERL_NIF_TERM out;	\
		unsigned char *buf = enif_make_new_binary(env, sizeof(sha3_##bits##_ctx_t), &out);	\
		struct sha3_##bits##_ctx_s *sponge = (struct sha3_##bits##_ctx_s *)(buf);	\
	\
		(void) sha3_##bits##_init(sponge);	\
	\
		return enif_make_tuple2(env, ATOM_sha3_##bits, out);	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_sha3_##bits##_update_nif_2(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		int arity;	\
		const ERL_NIF_TERM *state;	\
		ErlNifBinary state_bin;	\
		ErlNifBinary in;	\
		\
		if (argc != 2 || !enif_get_tuple(env, argv[0], &arity, &state)	\
			|| arity != 2	\
			|| state[0] != ATOM_sha3_##bits	\
			|| !enif_inspect_binary(env, state[1], &state_bin)	\
			|| state_bin.size != sizeof(sha3_##bits##_ctx_t)	\
			|| !enif_inspect_binary(env, argv[1], &in)) {	\
			return enif_make_badarg(env);	\
		}	\
		\
		if (in.size <= MAX_PER_SLICE) {	\
			ERL_NIF_TERM out;	\
			unsigned char *buf = enif_make_new_binary(env, state_bin.size, &out);	\
			(void) memcpy(buf, state_bin.data, state_bin.size);	\
			struct sha3_##bits##_ctx_s *sponge = (struct sha3_##bits##_ctx_s *)(buf);	\
		\
			(void) sha3_##bits##_update(sponge, in.data, in.size);	\
		\
			return enif_make_tuple2(env, ATOM_sha3_##bits, out);	\
		}	\
		\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource = enif_alloc_resource(resource_type, state_bin.size);	\
		(void) memcpy(resource, state_bin.data, state_bin.size);	\
		\
		ERL_NIF_TERM newargv[4];	\
		\
		newargv[0] = argv[1];					/* In */	\
		newargv[1] = enif_make_ulong(env, MAX_PER_SLICE);	/* MaxPerSlice */	\
		newargv[2] = enif_make_ulong(env, 0);			/* Offset */	\
		newargv[3] = enif_make_resource(env, resource);		/* Sponge */	\
		\
		(void) enif_release_resource(resource);	\
		\
		return enif_schedule_nif(env, "sha3_" #bits "_update", 0, keccakf1600_sha3_##bits##_update_nif_4, 4, newargv);	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_sha3_##bits##_update_nif_4(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		ErlNifBinary in;	\
		unsigned long max_per_slice;	\
		unsigned long offset;	\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource;	\
		\
		if (argc != 4 || !enif_inspect_binary(env, argv[0], &in)	\
				|| !enif_get_ulong(env, argv[1], &max_per_slice)	\
				|| !enif_get_ulong(env, argv[2], &offset)	\
				|| !enif_get_resource(env, argv[3], resource_type, &resource)) {	\
			return enif_make_badarg(env);	\
		}	\
		\
		struct sha3_##bits##_ctx_s *sponge = (struct sha3_##bits##_ctx_s *)(resource);	\
		\
		struct timeval start;	\
		struct timeval stop;	\
		struct timeval slice;	\
		unsigned long end;	\
		unsigned long i;	\
		int percent;	\
		int total = 0;	\
		\
		end = offset + max_per_slice;	\
		\
		if (end > in.size) {	\
			end = in.size;	\
		}	\
		\
		i = offset;	\
		\
		while (i < in.size) {	\
			(void) gettimeofday(&start, NULL);	\
			(void) sha3_##bits##_update(sponge, (uint8_t *)(in.data) + i, end - i);	\
			i = end;	\
			if (i == in.size) {	\
				break;	\
			}	\
			(void) gettimeofday(&stop, NULL);	\
			/* determine how much of the timeslice was used */	\
			timersub(&stop, &start, &slice);	\
			percent = (int)((slice.tv_sec*1000000+slice.tv_usec)/10);	\
			total += percent;	\
			if (percent > 100) {	\
				percent = 100;	\
			} else if (percent == 0) {	\
				percent = 1;	\
			}	\
			if (enif_consume_timeslice(env, percent)) {	\
				/* the timeslice has been used up, so adjust our max_per_slice byte count based on the processing we've done, then reschedule to run again */	\
				max_per_slice = i - offset;	\
				if (total > 100) {	\
					int m = (int)(total/100);	\
					if (m == 1) {	\
						max_per_slice -= (unsigned long)(max_per_slice*(total-100)/100);	\
					} else {	\
						max_per_slice = (unsigned long)(max_per_slice/m);	\
					}	\
				}	\
				ERL_NIF_TERM newargv[4];	\
				newargv[0] = argv[0];					/* In */	\
				newargv[1] = enif_make_ulong(env, max_per_slice);	/* MaxPerSlice */	\
				newargv[2] = enif_make_ulong(env, i);			/* Offset */	\
				newargv[3] = argv[3];					/* Sponge */	\
				return enif_schedule_nif(env, "sha3_" #bits "_update", 0, keccakf1600_sha3_##bits##_update_nif_4, argc, newargv);	\
			}	\
			end += max_per_slice;	\
			if (end > in.size) {	\
				end = in.size;	\
			}	\
		}	\
		\
		ERL_NIF_TERM out;	\
		unsigned char *buf = enif_make_new_binary(env, sizeof(sha3_##bits##_ctx_t), &out);	\
		(void) memcpy(buf, resource, sizeof(sha3_##bits##_ctx_t));	\
		\
		return enif_make_tuple2(env, ATOM_sha3_##bits, out);	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_sha3_##bits##_final_nif_1(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		int arity;	\
		const ERL_NIF_TERM *state;	\
		ErlNifBinary state_bin;	\
		\
		if (argc != 1 || !enif_get_tuple(env, argv[0], &arity, &state)	\
			|| arity != 2	\
			|| state[0] != ATOM_sha3_##bits	\
			|| !enif_inspect_binary(env, state[1], &state_bin)	\
			|| state_bin.size != sizeof(sha3_##bits##_ctx_t)) {	\
			return enif_make_badarg(env);	\
		}	\
		\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource = enif_alloc_resource(resource_type, state_bin.size);	\
		(void) memcpy(resource, state_bin.data, state_bin.size);	\
		struct sha3_##bits##_ctx_s *sponge = (struct sha3_##bits##_ctx_s *)(resource);	\
		ERL_NIF_TERM out;	\
		unsigned char *buf = enif_make_new_binary(env, bytes, &out);	\
		\
		(void) sha3_##bits##_final(sponge, buf, bytes);	\
		(void) sha3_##bits##_destroy(sponge);	\
		(void) enif_release_resource(resource);	\
		\
		return out;	\
	}

#define SHAKE_NIF(bits)	\
	static ERL_NIF_TERM	\
	keccakf1600_shake##bits##_nif_2(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		ErlNifBinary in;	\
		unsigned long outlen;	\
	\
		if (argc != 2 || !enif_inspect_binary(env, argv[0], &in)	\
				|| !enif_get_ulong(env, argv[1], &outlen)) {	\
			return enif_make_badarg(env);	\
		}	\
	\
		if (in.size <= MAX_PER_SLICE) {	\
			ERL_NIF_TERM out;	\
			unsigned char *buf = enif_make_new_binary(env, outlen, &out);	\
	\
			(void) shake##bits##_hash(buf, outlen, in.data, in.size);	\
	\
			return out;	\
		}	\
	\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource = enif_alloc_resource(resource_type, sizeof(shake##bits##_ctx_t));	\
		struct shake##bits##_ctx_s *sponge = (struct shake##bits##_ctx_s *)(resource);	\
		(void) shake##bits##_init(sponge);	\
	\
		ERL_NIF_TERM newargv[5];	\
	\
		newargv[0] = argv[0];					/* In */	\
		newargv[1] = argv[1];					/* Outlen */	\
		newargv[2] = enif_make_ulong(env, MAX_PER_SLICE);	/* MaxPerSlice */	\
		newargv[3] = enif_make_ulong(env, 0);			/* Offset */	\
		newargv[4] = enif_make_resource(env, resource);		/* Sponge */	\
	\
		(void) enif_release_resource(resource);	\
	\
		return enif_schedule_nif(env, "shake" #bits, 0, keccakf1600_shake##bits##_nif_5, 5, newargv);	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_shake##bits##_nif_5(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		ErlNifBinary in;	\
		unsigned long outlen;	\
		unsigned long max_per_slice;	\
		unsigned long offset;	\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource;	\
	\
		if (argc != 5 || !enif_inspect_binary(env, argv[0], &in)	\
				|| !enif_get_ulong(env, argv[1], &outlen)	\
				|| !enif_get_ulong(env, argv[2], &max_per_slice)	\
				|| !enif_get_ulong(env, argv[3], &offset)	\
				|| !enif_get_resource(env, argv[4], resource_type, &resource)) {	\
			return enif_make_badarg(env);	\
		}	\
	\
		struct shake##bits##_ctx_s *sponge = (struct shake##bits##_ctx_s *)(resource);	\
	\
		struct timeval start;	\
		struct timeval stop;	\
		struct timeval slice;	\
		unsigned long end;	\
		unsigned long i;	\
		int percent;	\
		int total = 0;	\
	\
		end = offset + max_per_slice;	\
	\
		if (end > in.size) {	\
			end = in.size;	\
		}	\
	\
		i = offset;	\
	\
		while (i < in.size) {	\
			(void) gettimeofday(&start, NULL);	\
			(void) shake##bits##_update(sponge, (uint8_t *)(in.data) + i, end - i);	\
			i = end;	\
			if (i == in.size) {	\
				break;	\
			}	\
			(void) gettimeofday(&stop, NULL);	\
			/* determine how much of the timeslice was used */	\
			timersub(&stop, &start, &slice);	\
			percent = (int)((slice.tv_sec*1000000+slice.tv_usec)/10);	\
			total += percent;	\
			if (percent > 100) {	\
				percent = 100;	\
			} else if (percent == 0) {	\
				percent = 1;	\
			}	\
			if (enif_consume_timeslice(env, percent)) {	\
				/* the timeslice has been used up, so adjust our max_per_slice byte count based on the processing we've done, then reschedule to run again */	\
				max_per_slice = i - offset;	\
				if (total > 100) {	\
					int m = (int)(total/100);	\
					if (m == 1) {	\
						max_per_slice -= (unsigned long)(max_per_slice*(total-100)/100);	\
					} else {	\
						max_per_slice = (unsigned long)(max_per_slice/m);	\
					}	\
				}	\
				ERL_NIF_TERM newargv[5];	\
				newargv[0] = argv[0];					/* In */	\
				newargv[1] = argv[1];					/* Outlen */	\
				newargv[2] = enif_make_ulong(env, max_per_slice);	/* MaxPerSlice */	\
				newargv[3] = enif_make_ulong(env, i);			/* Offset */	\
				newargv[4] = argv[4];					/* Sponge */	\
				return enif_schedule_nif(env, "shake" #bits, 0, keccakf1600_shake##bits##_nif_5, argc, newargv);	\
			}	\
			end += max_per_slice;	\
			if (end > in.size) {	\
				end = in.size;	\
			}	\
		}	\
	\
		ERL_NIF_TERM out;	\
		unsigned char *buf = enif_make_new_binary(env, outlen, &out);	\
	\
		(void) shake##bits##_final(sponge, buf, outlen);	\
		(void) shake##bits##_destroy(sponge);	\
	\
		return out;	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_shake##bits##_init_nif_0(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		if (argc != 0) {	\
			return enif_make_badarg(env);	\
		}	\
	\
		ERL_NIF_TERM out;	\
		unsigned char *buf = enif_make_new_binary(env, sizeof(shake##bits##_ctx_t), &out);	\
		struct shake##bits##_ctx_s *sponge = (struct shake##bits##_ctx_s *)(buf);	\
	\
		(void) shake##bits##_init(sponge);	\
	\
		return enif_make_tuple2(env, ATOM_shake##bits, out);	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_shake##bits##_update_nif_2(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		int arity;	\
		const ERL_NIF_TERM *state;	\
		ErlNifBinary state_bin;	\
		ErlNifBinary in;	\
		\
		if (argc != 2 || !enif_get_tuple(env, argv[0], &arity, &state)	\
			|| arity != 2	\
			|| state[0] != ATOM_shake##bits	\
			|| !enif_inspect_binary(env, state[1], &state_bin)	\
			|| state_bin.size != sizeof(shake##bits##_ctx_t)	\
			|| !enif_inspect_binary(env, argv[1], &in)) {	\
			return enif_make_badarg(env);	\
		}	\
		\
		if (in.size <= MAX_PER_SLICE) {	\
			ERL_NIF_TERM out;	\
			unsigned char *buf = enif_make_new_binary(env, state_bin.size, &out);	\
			(void) memcpy(buf, state_bin.data, state_bin.size);	\
			struct shake##bits##_ctx_s *sponge = (struct shake##bits##_ctx_s *)(buf);	\
		\
			(void) shake##bits##_update(sponge, in.data, in.size);	\
		\
			return enif_make_tuple2(env, ATOM_shake##bits, out);	\
		}	\
		\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource = enif_alloc_resource(resource_type, state_bin.size);	\
		(void) memcpy(resource, state_bin.data, state_bin.size);	\
		\
		ERL_NIF_TERM newargv[4];	\
		\
		newargv[0] = argv[1];					/* In */	\
		newargv[1] = enif_make_ulong(env, MAX_PER_SLICE);	/* MaxPerSlice */	\
		newargv[2] = enif_make_ulong(env, 0);			/* Offset */	\
		newargv[3] = enif_make_resource(env, resource);		/* Sponge */	\
		\
		(void) enif_release_resource(resource);	\
		\
		return enif_schedule_nif(env, "shake" #bits "_update", 0, keccakf1600_shake##bits##_update_nif_4, 4, newargv);	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_shake##bits##_update_nif_4(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		ErlNifBinary in;	\
		unsigned long max_per_slice;	\
		unsigned long offset;	\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource;	\
		\
		if (argc != 4 || !enif_inspect_binary(env, argv[0], &in)	\
				|| !enif_get_ulong(env, argv[1], &max_per_slice)	\
				|| !enif_get_ulong(env, argv[2], &offset)	\
				|| !enif_get_resource(env, argv[3], resource_type, &resource)) {	\
			return enif_make_badarg(env);	\
		}	\
		\
		struct shake##bits##_ctx_s *sponge = (struct shake##bits##_ctx_s *)(resource);	\
		\
		struct timeval start;	\
		struct timeval stop;	\
		struct timeval slice;	\
		unsigned long end;	\
		unsigned long i;	\
		int percent;	\
		int total = 0;	\
		\
		end = offset + max_per_slice;	\
		\
		if (end > in.size) {	\
			end = in.size;	\
		}	\
		\
		i = offset;	\
		\
		while (i < in.size) {	\
			(void) gettimeofday(&start, NULL);	\
			(void) shake##bits##_update(sponge, (uint8_t *)(in.data) + i, end - i);	\
			i = end;	\
			if (i == in.size) {	\
				break;	\
			}	\
			(void) gettimeofday(&stop, NULL);	\
			/* determine how much of the timeslice was used */	\
			timersub(&stop, &start, &slice);	\
			percent = (int)((slice.tv_sec*1000000+slice.tv_usec)/10);	\
			total += percent;	\
			if (percent > 100) {	\
				percent = 100;	\
			} else if (percent == 0) {	\
				percent = 1;	\
			}	\
			if (enif_consume_timeslice(env, percent)) {	\
				/* the timeslice has been used up, so adjust our max_per_slice byte count based on the processing we've done, then reschedule to run again */	\
				max_per_slice = i - offset;	\
				if (total > 100) {	\
					int m = (int)(total/100);	\
					if (m == 1) {	\
						max_per_slice -= (unsigned long)(max_per_slice*(total-100)/100);	\
					} else {	\
						max_per_slice = (unsigned long)(max_per_slice/m);	\
					}	\
				}	\
				ERL_NIF_TERM newargv[4];	\
				newargv[0] = argv[0];					/* In */	\
				newargv[1] = enif_make_ulong(env, max_per_slice);	/* MaxPerSlice */	\
				newargv[2] = enif_make_ulong(env, i);			/* Offset */	\
				newargv[3] = argv[3];					/* Sponge */	\
				return enif_schedule_nif(env, "shake" #bits "_update", 0, keccakf1600_shake##bits##_update_nif_4, argc, newargv);	\
			}	\
			end += max_per_slice;	\
			if (end > in.size) {	\
				end = in.size;	\
			}	\
		}	\
		\
		ERL_NIF_TERM out;	\
		unsigned char *buf = enif_make_new_binary(env, sizeof(shake##bits##_ctx_t), &out);	\
		(void) memcpy(buf, resource, sizeof(shake##bits##_ctx_t));	\
		\
		return enif_make_tuple2(env, ATOM_shake##bits, out);	\
	}	\
	\
	static ERL_NIF_TERM	\
	keccakf1600_shake##bits##_final_nif_2(ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])	\
	{	\
		int arity;	\
		const ERL_NIF_TERM *state;	\
		ErlNifBinary state_bin;	\
		unsigned long outlen;	\
		\
		if (argc != 2 || !enif_get_tuple(env, argv[0], &arity, &state)	\
			|| arity != 2	\
			|| state[0] != ATOM_shake##bits	\
			|| !enif_inspect_binary(env, state[1], &state_bin)	\
			|| state_bin.size != sizeof(shake##bits##_ctx_t)	\
			|| !enif_get_ulong(env, argv[1], &outlen)) {	\
			return enif_make_badarg(env);	\
		}	\
		\
		ErlNifResourceType *resource_type = (ErlNifResourceType *)(enif_priv_data(env));	\
		void *resource = enif_alloc_resource(resource_type, state_bin.size);	\
		(void) memcpy(resource, state_bin.data, state_bin.size);	\
		struct shake##bits##_ctx_s *sponge = (struct shake##bits##_ctx_s *)(resource);	\
		ERL_NIF_TERM out;	\
		unsigned char *buf = enif_make_new_binary(env, outlen, &out);	\
		\
		(void) shake##bits##_final(sponge, buf, outlen);	\
		(void) shake##bits##_destroy(sponge);	\
		(void) enif_release_resource(resource);	\
		\
		return out;	\
	}

SHA3_NIF(224, 28);
SHA3_NIF(256, 32);
SHA3_NIF(384, 48);
SHA3_NIF(512, 64);
SHAKE_NIF(128);
SHAKE_NIF(256);

#undef SHA3_NIF
#undef SHAKE_NIF

/*
 * Erlang NIF callbacks
 */
static int
keccakf1600_nif_load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info)
{
	/* Initialize common atoms */
	#define ATOM(Id, Value) { Id = enif_make_atom(env, Value); }
		ATOM(ATOM_sha3_224, "sha3_224");
		ATOM(ATOM_sha3_256, "sha3_256");
		ATOM(ATOM_sha3_384, "sha3_384");
		ATOM(ATOM_sha3_512, "sha3_512");
		ATOM(ATOM_shake128, "shake128");
		ATOM(ATOM_shake256, "shake256");
	#undef ATOM

	*priv_data = enif_open_resource_type(env, NULL, "keccakf1600_sponge", NULL, ERL_NIF_RT_CREATE | ERL_NIF_RT_TAKEOVER, NULL);
	return 0;
}

static int
keccakf1600_nif_upgrade(ErlNifEnv *env, void **priv_data, void **old_priv_data, ERL_NIF_TERM load_info)
{
	*priv_data = enif_open_resource_type(env, NULL, "keccakf1600_sponge", NULL, ERL_NIF_RT_TAKEOVER, NULL);
	return 0;
}

ERL_NIF_INIT(keccakf1600, keccakf1600_nif_funcs, keccakf1600_nif_load, NULL, keccakf1600_nif_upgrade, NULL);
